### ISSUE TITLE
EventDataDeserializationException: Failed to deserialize data of EventHeaderV4

### ISSUE THREAD

**Issue by jkrcma** [zendesk/maxwell#1133]
Followup of #1124, I found out the exception in Maxwell logs:

```
08:05:05,542 WARN  BinlogConnectorLifecycleListener - Communication failure.
com.github.shyiko.mysql.binlog.event.deserialization.EventDataDeserializationException: Failed to deserialize data of EventHeaderV4{timestamp=1539676889000, eventType=EXT_UPDATE_ROWS, serverId=4, headerLength=19, dataLength=631, nextPosition=487575926, flags=0}
        at com.github.shyiko.mysql.binlog.event.deserialization.EventDeserializer.deserializeEventData(EventDeserializer.java:233) ~[mysql-binlog-connector-java-0.16.1.jar:0.16.1]
        at com.github.shyiko.mysql.binlog.event.deserialization.EventDeserializer.nextEvent(EventDeserializer.java:201) ~[mysql-binlog-connector-java-0.16.1.jar:0.16.1]
        at com.github.shyiko.mysql.binlog.BinaryLogClient.listenForEventPackets(BinaryLogClient.java:890) [mysql-binlog-connector-java-0.16.1.jar:0.16.1]
        at com.github.shyiko.mysql.binlog.BinaryLogClient.connect(BinaryLogClient.java:559) [mysql-binlog-connector-java-0.16.1.jar:0.16.1]
        at com.github.shyiko.mysql.binlog.BinaryLogClient$7.run(BinaryLogClient.java:793) [mysql-binlog-connector-java-0.16.1.jar:0.16.1]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_181]
Caused by: java.io.EOFException
        at com.github.shyiko.mysql.binlog.io.ByteArrayInputStream.read(ByteArrayInputStream.java:190) ~[mysql-binlog-connector-java-0.16.1.jar:0.16.1]
        at java.io.InputStream.read(InputStream.java:170) ~[?:1.8.0_181]
        at com.github.shyiko.mysql.binlog.io.ByteArrayInputStream.fill(ByteArrayInputStream.java:96) ~[mysql-binlog-connector-java-0.16.1.jar:0.16.1]
        at com.github.shyiko.mysql.binlog.io.ByteArrayInputStream.read(ByteArrayInputStream.java:89) ~[mysql-binlog-connector-java-0.16.1.jar:0.16.1]
        at com.github.shyiko.mysql.binlog.event.deserialization.AbstractRowsEventDataDeserializer.deserializeVarString(AbstractRowsEventDataDeserializer.java:376) ~[mysql-binlog-connector-java-0.16.1.jar:0.16.1]
        at com.github.shyiko.mysql.binlog.event.deserialization.AbstractRowsEventDataDeserializer.deserializeCell(AbstractRowsEventDataDeserializer.java:184) ~[mysql-binlog-connector-java-0.16.1.jar:0.16.1]
        at com.github.shyiko.mysql.binlog.event.deserialization.AbstractRowsEventDataDeserializer.deserializeRow(AbstractRowsEventDataDeserializer.java:138) ~[mysql-binlog-connector-java-0.16.1.jar:0.16.1]
        at com.github.shyiko.mysql.binlog.event.deserialization.UpdateRowsEventDataDeserializer.deserializeRows(UpdateRowsEventDataDeserializer.java:72) ~[mysql-binlog-connector-java-0.16.1.jar:0.16.1]
        at com.github.shyiko.mysql.binlog.event.deserialization.UpdateRowsEventDataDeserializer.deserialize(UpdateRowsEventDataDeserializer.java:58) ~[mysql-binlog-connector-java-0.16.1.jar:0.16.1]
        at com.github.shyiko.mysql.binlog.event.deserialization.UpdateRowsEventDataDeserializer.deserialize(UpdateRowsEventDataDeserializer.java:33) ~[mysql-binlog-connector-java-0.16.1.jar:0.16.1]
        at com.github.shyiko.mysql.binlog.event.deserialization.EventDeserializer.deserializeEventData(EventDeserializer.java:227) ~[mysql-binlog-connector-java-0.16.1.jar:0.16.1]
        ... 5 more
08:05:05,549 INFO  BinlogConnectorLifecycleListener - Binlog disconnected.
08:05:05,641 WARN  BinlogConnectorReplicator - replicator stopped at position: 00003306-1111-1111-1111-111111111111:1-16698080696,00003306-2222-2222-2222-222222222222:1-10138717575 -- restarting
```

After the replicator is restarted Maxwell falls into a dead loop immediately:

```
com.github.shyiko.mysql.binlog.network.ServerException: could not find next log; the first event 'vilik-bin.022246' at 487575276, the last event read from '/data2/mysql_binlog/vilik-bin.022246' at 487575926, the last byte read from '/data2/mysql_binlog/vilik-bin.022246' at 487575926.
        at com.github.shyiko.mysql.binlog.BinaryLogClient.listenForEventPackets(BinaryLogClient.java:882) [mysql-binlog-connector-java-0.16.1.jar:0.16.1]
        at com.github.shyiko.mysql.binlog.BinaryLogClient.connect(BinaryLogClient.java:559) [mysql-binlog-connector-java-0.16.1.jar:0.16.1]
        at com.github.shyiko.mysql.binlog.BinaryLogClient$7.run(BinaryLogClient.java:793) [mysql-binlog-connector-java-0.16.1.jar:0.16.1]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_181]
08:05:05,770 INFO  BinlogConnectorLifecycleListener - Binlog disconnected.
08:05:05,868 WARN  BinlogConnectorReplicator - replicator stopped at position: 00003306-1111-1111-1111-111111111111:1-16698080696,00003306-2222-2222-2222-222222222222:1-10138717575 -- restarting
```

Output of trx `10138717575` from binlog (shortened):

```
# at 1015608696
#181016 10:01:29 server id 4  end_log_pos 1015609390 CRC32 0x3625e2fb   Update_rows: table id 255
# at 1015609390
#181016 10:01:29 server id 4  end_log_pos 1015609443 CRC32 0xd5dc98b3   Write_rows: table id 321
# at 1015609443
#181016 10:01:29 server id 4  end_log_pos 1015610089 CRC32 0x75d5e1b0   Update_rows: table id 255
# at 1015610089
#181016 10:01:29 server id 4  end_log_pos 1015610142 CRC32 0xe18a5d14   Write_rows: table id 321

... ~150 times ...

# at 1015677435
#181016 10:01:29 server id 4  end_log_pos 1015677488 CRC32 0xb1d0ec02   Write_rows: table id 321 flags: STMT_END_F
```

Output of trx `10138717576` from binlog:

```
# at 1015682189
#181016 10:01:29 server id 4  end_log_pos 1015682255 CRC32 0x12ec29c1   Write_rows: table id 718 flags: STMT_END_F
```

I noticed that the exception contains `eventType=EXT_UPDATE_ROWS`, however trx `10138717576` contains just a single `INSERT`, however lot of `INSERT`s and `UPDATE`s are present in the previous transaction which is the last one persisted in GTID range.

We do normally execute big transactions like the one above and Maxwell replicator crashes just rarely. Restarting Maxwell fixes the issue.

**Comment by ericwush** [comment]
@jkrcma do you mean after getting `com.github.shyiko.mysql.binlog.network.ServerException` maxwell was still running? From code it seems maxwell should crash.

**Comment by jkrcma** [comment]
@ericwush It was still running and cycling this error 2-3 times per second. Sometimes it just recovers from error state by its own, but without any reasonable log message.

**Comment by osheroff** [comment]
i was able to repro this in my tests; binlog connector was trying to point at some spot inside the GTID transaction and mysql wasn't having it.  I've fixed this as part of #1129, closing down the dupe now.
